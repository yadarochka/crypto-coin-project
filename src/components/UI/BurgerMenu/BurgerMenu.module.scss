@import "~../../../styles/variables.scss";
$burger-width: 20px;

.btn {
  width: $burger-width;
  z-index: 101;
  height: 20px;
  position: relative;
  cursor: pointer;
  & span {
    transition: all 0.3s ease-in-out;
  }
}

.top {
  position: absolute;
  top: 0;
  width: $burger-width;
  background-color: black;
  height: 2px;
}
.bottom {
  position: absolute;
  bottom: 0;
  width: $burger-width;
  background-color: black;
  height: 2px;
}
.middle {
  position: absolute;
  bottom: 9px;
  width: $burger-width;
  background-color: black;
  height: 2px;
}

.menu {
  position: fixed;
  width: 100vw;
  height: 100vh;
  top: 0;
  right: 0;
  z-index: 100;
}

.blur {
  position: absolute;
  width: 100%;
  height: 100vh;
  right: 29%;
  backdrop-filter: blur(2px);
}

.list {
  position: absolute;
  display: flex;
  flex-direction: column;
  background-color: white;
  top: 0;
  right: -30%;
  width: 30%;
  height: 100%;
  padding: 1rem 1rem;
  gap: 5px;
  transition: right 0.5s ease;
}

li {
  text-decoration: none;
  list-style: none;
}

a {
  text-decoration: none;
  color: $black-color;
}

.menuActive {
  right: 0;
}

.btnActive {
  &.top {
    transform: rotate(-45deg);
    margin-top: 7px;
  }
  &.bottom {
    opacity: 0;
    transform: rotate(45deg);
  }
  &.middle {
    transform: rotate(45deg);
    margin-bottom: 2px;
  }
}
