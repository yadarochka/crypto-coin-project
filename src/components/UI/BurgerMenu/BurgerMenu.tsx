import classnames from "classnames";

import React, { FC, useState } from "react";
import { NavLink } from "react-router-dom";

import { useKeyboardEvent } from "utils/useKeyboardEvent";

import styles from "./BurgerMenu.module.scss";

export type BurgerMenuItems = {
  linkTo: string;
  imgSrc?: any;
  label: string;
};

type BurgerMenuProps = {
  items: BurgerMenuItems[];
};

export const BurgerMenu: FC<BurgerMenuProps> = ({ items }) => {
  const [isActive, setActive] = useState<boolean>(false);
  const burgerIsVisibleChange = () => setActive(!isActive);

  useKeyboardEvent("Esc", burgerIsVisibleChange, "burgerMenu", [isActive]);
  useKeyboardEvent("Enter", burgerIsVisibleChange, "burgerMenu-icon", [
    isActive,
  ]);

  const Menu = () => (
    <div className={styles.menu} data-testid="menu">
      <div className={isActive ? styles.blur : ""} />
      <nav
        className={classnames(
          isActive ? styles.menuActive : styles.menuNoActive,
          styles.list
        )}
      >
        <h2 className={styles.title}>Pages</h2>
        {items.map((link) => (
          <li>
            <NavLink
              key={link.label}
              data-testId={`${link.label.toLocaleLowerCase()}-nav`}
              className={styles.item}
              to={link.linkTo}
              onClick={burgerIsVisibleChange}
            >
              <div className={styles.link}>
                <img
                  src={link.imgSrc}
                  alt={link.label}
                  className={styles.linkImg}
                />
                {link.label}
              </div>
            </NavLink>
          </li>
        ))}
      </nav>
    </div>
  );

  return (
    <>
      <div
        id="burgerMenu-icon"
        data-testId="burger-menu"
        tabIndex={0}
        className={classnames(styles.btn, { [styles.btnActive]: isActive })}
        onClick={burgerIsVisibleChange}
      >
        <span
          className={classnames(styles.top, { [styles.btnActive]: isActive })}
        />
        <span
          className={classnames(styles.middle, {
            [styles.btnActive]: isActive,
          })}
        />
        <span
          className={classnames(styles.bottom, {
            [styles.btnActive]: isActive,
          })}
        />
      </div>
      {isActive && <Menu />}
    </>
  );
};
