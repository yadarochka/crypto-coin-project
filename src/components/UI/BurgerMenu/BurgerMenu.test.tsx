import "@testing-library/jest-dom";
import { cleanup, fireEvent, render, screen } from "@testing-library/react";

import { MemoryRouter } from "react-router-dom";

import { BurgerMenu } from "./BurgerMenu";

const items = [
  {
    linkTo: "/abc",
    imgSrc: "",
    label: "ABC",
  },
  {
    linkTo: "/123",
    imgSrc: "",
    label: "123",
  },
  {
    linkTo: "/bitcoin",
    imgSrc: "",
    label: "Bitcoin",
  },
];

describe("Тест BurgerMenu", () => {
  it("Snapshot-тестирование бургер-иконки", () => {
    const { container } = render(<BurgerMenu items={items} />);

    expect(container).toMatchSnapshot();
  });

  it("Snapshot-тестирование меню", () => {
    render(
      <MemoryRouter>
        <BurgerMenu items={items} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByTestId("burger-menu"));
    const menu = screen.getByTestId("menu");

    expect(menu).toMatchSnapshot();
  });

  it("Бургер-иконка рендерится", () => {
    render(<BurgerMenu items={items} />);

    expect(screen.getByTestId("burger-menu")).toBeInTheDocument();
  });

  it("Меню не видно в выключенном состоянии", () => {
    render(<BurgerMenu items={items} />);

    expect(
      screen.queryByLabelText<HTMLLinkElement>(/ABC/)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByLabelText<HTMLLinkElement>(/123/)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByLabelText<HTMLLinkElement>(/Bitcoin/)
    ).not.toBeInTheDocument();
  });

  it("Меню появляется при первом нажатии на бургер-иконку и скрывается при повторном", () => {
    render(
      <MemoryRouter>
        <BurgerMenu items={items} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByTestId("burger-menu"));

    expect(screen.getByText<HTMLDivElement>(/ABC/)).toBeInTheDocument();
    expect(screen.getByText<HTMLDivElement>(/123/)).toBeInTheDocument();
    expect(screen.getByText<HTMLDivElement>(/Bitcoin/)).toBeInTheDocument();

    fireEvent.click(screen.getByTestId("burger-menu"));

    expect(
      screen.queryByLabelText<HTMLLinkElement>(/ABC/)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByLabelText<HTMLLinkElement>(/123/)
    ).not.toBeInTheDocument();
    expect(
      screen.queryByLabelText<HTMLLinkElement>(/Bitcoin/)
    ).not.toBeInTheDocument();
  });

  afterEach(cleanup);
});
