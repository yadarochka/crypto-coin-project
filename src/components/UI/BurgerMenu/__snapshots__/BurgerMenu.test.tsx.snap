// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест BurgerMenu Snapshot-тестирование бургер-иконки 1`] = `
<div>
  <div
    class=""
    data-testid="burger-menu"
    id="burgerMenu-icon"
    tabindex="0"
  >
    <span
      class=""
    />
    <span
      class=""
    />
    <span
      class=""
    />
  </div>
</div>
`;

exports[`Тест BurgerMenu Snapshot-тестирование меню 1`] = `
<div
  data-testid="menu"
>
  <div />
  <nav
    class=""
  >
    <h2>
      Pages
    </h2>
    <li>
      <a
        class=""
        data-testid="abc-nav"
        href="/abc"
      >
        <div>
          <img
            alt="ABC"
            src=""
          />
          ABC
        </div>
      </a>
    </li>
    <li>
      <a
        class=""
        data-testid="123-nav"
        href="/123"
      >
        <div>
          <img
            alt="123"
            src=""
          />
          123
        </div>
      </a>
    </li>
    <li>
      <a
        class=""
        data-testid="bitcoin-nav"
        href="/bitcoin"
      >
        <div>
          <img
            alt="Bitcoin"
            src=""
          />
          Bitcoin
        </div>
      </a>
    </li>
  </nav>
</div>
`;
